uncompress to floating-point if justified by BSCALE,BZERO/dithering

PaulPrice opened the pull request:
This bugfix is required to read certain compressed images correctly.

I would request that it be targeted for 2.0.2.

pllim commented:
"Reviewers" drop-down menu is not working for me. c/c @MSeifert04 and @saimn 

(@drdavella too in case you are interested to have a look since you dabbled in compressed stuff recently)

MSeifert04 commented on astropy/io/fits/src/compressionmodule.c:
```diff
@@ -1061,6 +1061,24 @@ PyObject* compression_decompress_hdu(PyObject* self, PyObject* args)
         return NULL;
     }
+    // non-integer scaling or subtractive dithering means floating-point
+    if ((long)fileptr->Fptr->cn_bscale != fileptr->Fptr->cn_bscale ||
```
Is that how one tests for floats in `C`? It's probably correct but it seems weird.

PaulPrice replied:
Should be.  The same implementation is used in cfitsio.

MSeifert04 replied:
okay, I was thinking about possible overflow when casting `float` to `signed long integer`. I never know if that's just implementation defined or undefined.

MSeifert04 replied:
I'm not a C programmer so if you say that's okay I'll trust you :)

MSeifert04 commented on astropy/io/fits/tests/test_image.py:
```diff
@@ -1777,6 +1778,37 @@ def test_compressed_header_missing_znaxis(self):
 
+    @pytest.mark.parametrize(('keyword', 'dtype', 'expected'),
+                             [(kk, dd, ee) for kk, (dd, ee) in
+                             itertools.product(["BSCALE", "BZERO"],
```
These are only 8 cases currently. I agree that `product` and `zip` are nice but I had to run this in the interpreter to work out what it evaluated to. Maybe that could be spelled out explicitly?

```
    @pytest.mark.parametrize(
        ('keyword', 'dtype', 'expected'),
        [('BSCALE', (numpy.uint8, numpy.float32), ('BSCALE', (numpy.int16, numpy.float32),
         ('BSCALE', (numpy.int32, numpy.float64), ('BZERO', (numpy.uint8, numpy.float32),
         ('BZERO', (numpy.int16, numpy.float32), ('BZERO', numpy.int32, numpy.float64)]
```

MSeifert04 commented on astropy/io/fits/tests/test_image.py:
```diff
@@ -1777,6 +1778,37 @@ def test_compressed_header_missing_znaxis(self):
+        hdu = fits.CompImageHDU(np.arange(0, 10, dtype=dtype))
+        hdu.header[keyword] = value
+        hdu.writeto(self.temp('test.fits'))
+        del hdu
```
is there a reason why you have to delete it here? Just curious. :)

PaulPrice replied:
I use `hdu` in the next line, and just wanted to be clear it's not this `hdu`.

saimn commented:
@bsipocz - There is a test failure on Py3.4, maybe related to your recent PR ? 
`pkg_resources.VersionConflict: (py 1.4.32 (/home/travis/miniconda/envs/test/lib/python3.4/site-packages), Requirement.parse('py>=1.4.33'))`

@PaulPrice - A rebase is needed to fix the changelog conflict with your other PR, sorry about that.

saimn commented on astropy/io/fits/tests/test_image.py:
```diff
@@ -1778,6 +1778,38 @@ def test_compressed_header_missing_znaxis(self):
+        If BSCALE,BZERO is set to floating point values, the image
+        should be floating-point.
+
+        Second part of https://github.com/astropy/astropy/pull/6486
```
The sentence should be updated.

bsipocz commented:
@saimn - actually this is very weird, looks like some the changes in that PR is seen, but not all. I think a rebase will solve the failure.

PaulPrice commented:
Rebased against master and updated that docstring.

PaulPrice commented:
Green lights, ready to merge!

MSeifert04 commented on astropy/io/fits/src/compressionmodule.c:
```diff
@@ -1063,6 +1063,24 @@ PyObject* compression_decompress_hdu(PyObject* self, PyObject* args)
         return NULL;
     }
+    // non-integer scaling or subtractive dithering means floating-point
+    if (floor(fileptr->Fptr->cn_bscale) != fileptr->Fptr->cn_bscale ||
```
Yes, I think that's somewhat better although I'm still not convinced that this is correct. It could be a float that is bigger than the largest 64bit integer and pass this test.

I'm really wondering: Is there no way to safely check if a `double` is also a valid `long` in C?

PaulPrice replied:
I remember that the same test in cfitsio is more involved than what I've written here, probably for the reason you've mentioned. I'll expand it.

MSeifert04 commented on astropy/io/fits/src/compressionmodule.c:
```diff
@@ -1063,6 +1063,24 @@ PyObject* compression_decompress_hdu(PyObject* self, PyObject* args)
+        floor(fileptr->Fptr->cn_bzero) != fileptr->Fptr->cn_bzero ||
+        fileptr->Fptr->
+#ifdef CFITSIO_SUPPORTS_SUBTRACTIVE_DITHER_2
+            quantize_method
```
personall I would prefer to put the whole statement in a `ifdef`, just so it's more readable:

```
#ifdef CFITSIO_SUPPORTS_SUBTRACTIVE_DITHER_2
            fileptr->Fptr->quantize_method != NO_DITHER
#else
            fileptr->Fptr->quantize_dither != NO_DITHER
#endif
```

MSeifert04 commented on astropy/io/fits/src/compressionmodule.c:
```diff
@@ -1063,6 +1063,24 @@ PyObject* compression_decompress_hdu(PyObject* self, PyObject* args)
+            quantize_dither
+#endif
+                != NO_DITHER) {
+        if (datatype == BYTE_IMG || datatype == SHORT_IMG) {
```
Is this some FITS convention? I could image a lot of (although artificial) cases where the result would be not representable with normal `float`s if I allow `BZERO` and `BSCALE` to be arbitary `double`s.

saimn replied:
I remember seeing the same conversion in the Python code, but I didn't find if it is described in the standard.

saimn replied:
There is https://github.com/astropy/astropy/blob/master/astropy/io/fits/hdu/image.py#L673 for instance, I was thinking of another piece of code but cannot find it right now.

MSeifert04 commented:
@saimn @pllim I'm not sure if it's actually worth pursuing my questions. So if any one you feel this PR is ready to be merged then please go ahead and merge it.

@PaulPrice Thank you for the patience and thanks for the pull request. I hope you don't mind all the questions.

saimn commented:
@MSeifert04 - No problem, it's better to double check now than having to debug C code later ;-).

saimn commented:
So this looks good to me, it can be merged once @MSeifert04 is happy with it :wink:

PaulPrice commented:
Thanks for pushing me on this @MSeifert04. This fixup commit converts to using cfitsio to determine the appropriate type. cfitsio has much more sophistication in [its type determination function](https://github.com/astropy/astropy/blob/master/cextern/cfitsio/lib/fitscore.c#L7345). I had tried this before, but it didn't work then; now it seems to be working.
I've left it as a "fixup" commit so you can clearly see what's changed. If you like it and it passes CI, we can squash it in before merging.

PaulPrice commented:
That's the failure I was seeing before. Funny it didn't show up when I ran the tests myself.
Investigating....

MSeifert04 commented on astropy/io/fits/src/compressionmodule.c:
```diff
@@ -1059,9 +1059,36 @@ PyObject* compression_decompress_hdu(PyObject* self, PyObject* args)
-    if (PyErr_Occurred()) {
-        return NULL;
+    fits_get_img_equivtype(fileptr, &datatype, &status);
+    if (status != 0) {
```
Is it possible that this should be `> 0` instead of `!= 0`? That's what the `ffgiet` function checks.

MSeifert04 replied:
That could also explain the `unknown error` problems.

PaulPrice commented:
Well, that seems to work. Let me know what you think, @MSeifert04.

MSeifert04 commented on astropy/io/fits/src/compressionmodule.c:
```diff
@@ -1016,6 +1016,96 @@ PyObject* compression_compress_hdu(PyObject* self, PyObject* args)
+        *imgtype = DOUBLE_IMG;
+        *numpytype = NPY_DOUBLE;
+    }
+    abort();  // programming error
```
Is it necessary to call `abort()` here? I mean the function would just end here anyway.

MSeifert04 commented on astropy/io/fits/src/compressionmodule.c:
```diff
@@ -1016,6 +1016,96 @@ PyObject* compression_compress_hdu(PyObject* self, PyObject* args)
+
+    switch (*imgtype)
+    {
+      case BYTE_IMG:   /* 8-bit image */
```
That's interesting. You called `bitpix_to_datatypes` which changes `imgtype` to `TBYTE`, etc. but it still works correctly with `case BYTE_IMG`?

PaulPrice replied:
Eeek, you're right.  I'll fix it.
Scary that it still worked............

MSeifert04 commented on astropy/io/fits/src/compressionmodule.c:
```diff
@@ -1016,6 +1016,96 @@ PyObject* compression_compress_hdu(PyObject* self, PyObject* args)
+        min_val = -2147483648.0;
+        max_val =  2147483647.0;
+        break;
+      default:  /* don't have to deal with other data types */
```
The `bitpix_to_datatypes` could also return `TLONGLONG`. That could warrant it's own case.

PaulPrice replied:
cfitsio won't compress 64-bit integers. I'll put in a trap for it though.

MSeifert04 commented on astropy/io/fits/src/compressionmodule.c:
```diff
@@ -1016,6 +1016,96 @@ PyObject* compression_compress_hdu(PyObject* self, PyObject* args)
+        *imgtype = SHORT_IMG;
+        *numpytype = NPY_INT16;
+    } else if ((min_val >= 0.0) && (max_val <= 65535.0)) {
+        *imgtype = USHORT_IMG;
```
I wonder if we should really support unsigned integers here. It seems like we haven't previously.

PaulPrice commented:
I'm so sorry, everyone! I don't think I'm normally this careless. Thanks again to @MSeifert04 for pointing out those problems. I've pushed some fixes.

PaulPrice commented:
Thanks @MSeifert04! Rebased against master, squashing the fixups.

MSeifert04 commented:
@PaulPrice Thank you! Unfortunately Travis CI seems to have missed the most recent commit.

Code-wise this looks good to me (one nitpick 😄 : The indentation seems off/random in a few places)!

PaulPrice commented:
I didn't think I did any bad indentation, but I'm not up on your coding standards. If you point out some specific instances, I'll fix them.

MSeifert04 commented:
I think the new code contains 2, 3 as well as 4 whitespace indentation levels. But that seems to be the case in a lot of functions of that file. And it's C code so why bother. If noone else chimes in I think you can leave it as-is.

PaulPrice commented:
2 spaces is standard for `switch`/`case` in coding standards I'm used to.
3 spaces is me failing to re-indent cfitsio code (it has horrible indentation practises). I identified one line that I missed, and have pushed a fixed version.
4 spaces is standard elsewhere.

bsipocz commented:
@MSeifert04 - If the style is not consistent anyway, I would suggest to fix everything in one big batch separately when/if addressing https://github.com/astropy/astropy/issues/2743

PaulPrice commented:
Travis isn't firing. Does it require an Admin to bump it, or can I just force-push an amended commit?

bsipocz commented:
A travis build seems to be queued to me here, even though github doesn't seem to know about it: https://travis-ci.org/astropy/astropy/builds/272733807

PaulPrice commented:
Coverage decreased, says the checks, presumably because I added more code. I'll see if I can add some more test cases that exercise the new code better.

PaulPrice commented:
Oh, no....

While investigating the coverage decrease, I added a test (appended below) that the code would produce integer types. That test fails, and it turns out that the cfitsio code I added isn't firing at all because the scaling (application of `BSCALE`, `BZERO`) has been explicitly disabled in the C so that it can be done in python (without any checking of types). Changing/fixing that is more than I want to bite off or feel comfortable doing (and I don't need it in order to complete this project for LSST). If I then rip out the (now apparently useless) cfitsio code I added (and @MSeifert04 so carefully vetted!), the tests that I had previously added to check that floating-point values are produced from integer images pass just fine. That is to say that this entire contribution can be boiled down to simply adding those tests --- no new active code.

I'm so sorry, everyone. I guess I made an error early on and then got tunnel vision, probably because after over a month of living and breathing FITS compression, I was desperate to get it all done. You can take it as a point of pride that your CI tools caught this and refused to let me have my stupid way. I'm sorry to have wasted your time. Thanks for your patience and support. I've been greatly impressed with the little glimpse I've had inside your project (both the people and technology).

Here's a test you might want to consider for the future (it currently fails; fixing that would bring you into closer conformity with cfitsio which may or may not be something you want):

    @pytest.mark.parametrize(
        ('dtype', 'bscale', 'bzero', 'expected'),
        [(np.uint8, 2.0, 0.0, np.int16), (np.uint8, 1.0, 1234.0, np.int16),
         (np.int16, 2.0, 0.0, np.int32), (np.int16, 1.0, 12345.0, np.int32),
         (np.int32, 2.0, 0.0, np.int64), (np.int32, 1.0, 123456.0, np.int64),
         (np.uint8, 256.0, 0.0, np.int32),
        ])
    def test_compressed_scaled_int(self, dtype, bscale, bzero, expected):
        """
        If BSCALE,BZERO is set to integer values, the image should be
        an integer sufficient to contain the range of possible values.

        Parameters
        ----------
        dtype : `numpy.dtype`
            Type of original array.
        bscale : `float`
            BSCALE to apply.
        bzero : `float`
            BZERO to apply.
        expected : `numpy.dtype`
            Expected type of uncompressed array.
        """
        hdu = fits.CompImageHDU(np.arange(0, 10, dtype=dtype))
        hdu.header["BSCALE"] = bscale
        hdu.header["BZERO"] = bzero
        hdu.writeto(self.temp('test.fits'))
        del hdu
        with fits.open(self.temp('test.fits')) as hdu:
            assert hdu[1].header["BSCALE"] == bscale
            assert hdu[1].header["BZERO"] == bzero
            assert hdu[1].data.dtype == expected

pllim commented:
> Oh, no....

IMHO, if this is merged with only the test, then a new issue should be opened to pursue the C route so it won't be forgotten.

MSeifert04 commented on CHANGES.rst:
```diff
@@ -300,6 +300,9 @@ astropy.io.fits
 - Support the ``GZIP_2`` FITS image compression algorithm as claimed
   in docs. [#6486]
 
+- Test that floating-point arrays are produced from compressed integer
```
Test changes generally don't need a changelog.

PaulPrice commented:
Travis failure is due to documentation/download failure:

    WARNING: failed to download http://maia.usno.navy.mil/ser7/finals2000A.all, using local IERS-B: <urlopen error [Errno -2] Name or service not known> [astropy.utils.iers.iers]
    WARNING: failed to download http://maia.usno.navy.mil/ser7/finals2000A.all, using local IERS-B: <urlopen error [Errno -2] Name or service not known>
looking for now-outdated files... none found

bsipocz commented:
Those warnings can be ignored. Thanks @PaulPrice!
